From 0aee2fdc55de2c9f4425af2f123fb1aa523f8ecb Mon Sep 17 00:00:00 2001
From: WANG Xuerui <xen0n@gentoo.org>
Date: Fri, 14 Feb 2025 16:59:24 +0800
Subject: [PATCH 8/9] Bug 1948198 - Enable LSX for libpng on LoongArch64.
 r?#gfx-reviewers

Enable LSX unconditionally for LoongArch64 because its availability
is guaranteed for LoongArch desktop and server platforms, which are
general-purpose and popular.

This change partially reverts D231469.

Differential Revision: https://phabricator.services.mozilla.com/D238166
---
 media/libpng/moz.build    | 7 ++++++-
 media/libpng/pnglibconf.h | 7 +++++++
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/media/libpng/moz.build b/media/libpng/moz.build
index 6fa28a814a757..9a860ae1ce8a5 100644
--- a/media/libpng/moz.build
+++ b/media/libpng/moz.build
@@ -66,7 +66,12 @@ if CONFIG['HAVE_ALTIVEC']:
     ]
 
 if CONFIG['TARGET_CPU'] == 'loongarch64':
-    CFLAGS += ['-mno-lsx']
+    DEFINES['MOZ_PNG_USE_LOONGARCH_LSX'] = True
+    CFLAGS += ['-mlsx']
+    UNIFIED_SOURCES += [
+        'loongarch/filter_lsx_intrinsics.c',
+        'loongarch/loongarch_lsx_init.c',
+    ]
 
 if CONFIG['MOZ_TREE_FREETYPE']:
     DEFINES['FT_CONFIG_OPTION_USE_PNG'] = True
diff --git a/media/libpng/pnglibconf.h b/media/libpng/pnglibconf.h
index 1cbb828436de2..8ef91386ffe43 100644
--- a/media/libpng/pnglibconf.h
+++ b/media/libpng/pnglibconf.h
@@ -54,6 +54,13 @@
 #  define PNG_ARM_NEON_OPT 0
 #endif
 
+#ifdef MOZ_PNG_USE_LOONGARCH_LSX
+#  undef PNG_LOONGARCH_LSX_OPT /* Let libpng decide */
+#  define PNG_ALIGNED_MEMORY_SUPPORTED
+#else
+#  define PNG_LOONGARCH_LSX_OPT 0
+#endif
+
 #ifdef MOZ_PNG_USE_MIPS_MSA
 #  undef PNG_MIPS_MSA_OPT
 #  define PNG_ALIGNED_MEMORY_SUPPORTED
-- 
2.48.1

